### PR TITLE
Streams

### PR DESCRIPTION
**Pull request by ianzapolsky**

### CODE REVIEW

**Comment by ianzapolsky** [comment]
tests seem to be failing because of travis weirdness. pull down from streams to take a look

**Comment by ianzapolsky** [comment]
take a look at the fibonacci syntax example. it's pretty ugly. one code objective for the next week could be to try to get this kind of operation to look better, from an aesthetic pov.

**Comment by ianzapolsky** [comment]
see [here](http://stackoverflow.com/questions/14531993/can-lambda-functions-be-recursive) for explanation of recursive lambda functions in c++

**Comment by natebrennand** [comment]
That's neat for recursive lambdas. I'm reading through this now and will try resolving travis

File: src/streams.h [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    from(T n, T step) {
+        return Stream<T>(n, [=]() -> Stream<T> {
+            return from(n + step, step);
+        });
+    }
+
+    // Construct a Stream, starting at n, incrementing by 1
+    template<typename T>
+    Stream<T>
+    from(T n) {
```
Comment (natebrennand) [inline src/streams.h]: we can skip this implementation if you set `step`'s default value to 1 in the definition.

File: src/streams.h [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        Stream<T> temp = tail();
+        return Stream<return_type>(func(Head), [temp, func]() -> Stream<return_type> {
+            return temp.map(func);
+        });
+    }
+    
+    // Construct a Stream, starting at n, incrementing by step
+    template<typename T>
+    Stream<T>
+    from(T n, T step) {
```
Comment (natebrennand) [inline src/streams.h]: like this:

```cpp
    template<typename T>
    Stream<T>
    from(T n, T step=1) {
        return Stream<T>(n, [=]() -> Stream<T> {
            return from(n + T(step), T(step));
        });
    }
```

**Comment by natebrennand** [comment]
this is super awesome :+1: 
it works for me locally, but let's try to get the build passing

**Comment by ianzapolsky** [comment]
@natebrennand any thoughts on a macro to ease the pain of writing a custom generator?

File: src/syntax/fibonacci.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "../cpp_collections.h"
+
+using namespace cpp_collections;
+
+int main() {
+
+    // Collection syntax
+
+    def_generator(fibs, int, int prev, int curr) {
```
Comment (natebrennand) [inline src/syntax/fibonacci.cpp]: Maybe not perfect, but I think this macro setup is plenty acceptable.

**Comment by natebrennand** [comment]
:+1: this all looks good to me. Messing around with the docker container now.

**Comment by ianzapolsky** [comment]
hey @natebrennand 

take a look at the 4 new types of stream generators i just added. `iterate` and `recurrence` are both really cool.

i'd like to get this branch merged asap. tutorial and readme will have to be updated. but this work makes that wacky macro pretty much moot, which is awesome.

**Comment by natebrennand** [comment]
So I'm honestly a bit confused by how the templates are working at this point but I definitely agree that this is a cleaner solution for the user than the macro.

**Pull request merged**